File: messagematch/src/main/java/org/forwoods/messagematch/match/JsonMatcher.java
package org.forwoods.messagematch.match;

import java.io.IOException;
import java.io.InputStream;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalTime;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.util.*;
import java.util.Map.Entry;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.node.NullNode;
import org.forwoods.messagematch.match.fieldmatchers.FieldMatcher;
import org.antlr.v4.runtime.BaseErrorListener;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.RecognitionException;
import org.antlr.v4.runtime.Recognizer;
import org.forwoods.messagematch.matchgrammar.MatcherLexer;
import org.forwoods.messagematch.matchgrammar.MatcherParser;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.node.ValueNode;
import org.forwoods.messagematch.util.PathExtractor;

import static com.fasterxml.jackson.databind.node.JsonNodeType.MISSING;
import static com.fasterxml.jackson.databind.node.JsonNodeType.NULL;

/**
 * @author Tom
 *
 */
public class JsonMatcher {
    static final ObjectMapper mapper = new ObjectMapper();

    private List<MatchError> errors = new ArrayList<>();

    private final JsonNode matcherNode;

    private final JsonNode concreteNode;

    private final Map<String, Object> bindings = new HashMap<>();

    protected long matchTime = -1;//can be overridden for unit tests of the matcher

    public JsonMatcher(InputStream matcher, InputStream concrete) throws IOException {
        this(readNodes(matcher), mapper.readTree(concrete));
    }

    public static JsonNode readNodes(InputStream matcher) throws IOException {
        return mapper.readTree(matcher);
    }

    public JsonMatcher(String matcher, String concrete) throws IOException {
        this(readNodes(matcher), readNodes(concrete));
    }

    public static JsonNode readNodes(String matcher) throws JsonProcessingException {
        return mapper.readTree(matcher);
    }

    public JsonMatcher(JsonNode matcher, String concrete) throws JsonProcessingException {
        this(matcher, readNodes(concrete));
    }

    public JsonMatcher(JsonNode matcher, JsonNode actual) {
        matcherNode = matcher;
        concreteNode = actual;
    }

    public enum MatcherOption{
        UNORDERED,
        STRICT
    }

    final Set<MatcherOption> options = new HashSet<>();

    boolean hasOption(MatcherOption option) {
        return options.contains(option);
    }

    public boolean matches(MatcherOption... options){
        Collections.addAll(this.options, options);
        return matches();
    }

    public boolean matches() {
        if (matchTime < 0) matchTime = System.currentTimeMillis();
        Instant i = Instant.ofEpochMilli(matchTime);
        ZonedDateTime t = ZonedDateTime.ofInstant(i, ZoneOffset.UTC);
        LocalDate d = t.toLocalDate();
        LocalTime lt = t.toLocalTime();
        bindings.put("date", d);
        bindings.put("time", lt);
        bindings.put("datetime", matchTime);

        return matches(new JsonPath("root", null), matcherNode, concreteNode);
    }

    private boolean matches(JsonPath path, JsonNode matcherNode, JsonNode concreteNode) {
        switch (matcherNode.getNodeType()) {

            case STRING:
            case BOOLEAN:
            case NUMBER:
                if (!(concreteNode instanceof ValueNode)) {
                    errors.add(new MatchError(path, "a value node", "A "+concreteNode.getNodeType() + " node"));
                    return false;
                }
                return matchPrimitive(path, (ValueNode) matcherNode, (ValueNode) concreteNode);

            case ARRAY:
                if (!(concreteNode instanceof ArrayNode)) {
                    errors.add(new MatchError(path, "an array node", "A "+concreteNode.getNodeType() + " node"));
                    return false;
                }
                return matchArray(path, (ArrayNode) matcherNode, (ArrayNode) concreteNode);
            case NULL:
                if (concreteNode.getNodeType()!=NULL) {
                    errors.add(new MatchError(path, "a null value", "a node with a value "+ concreteNode));
                    return false;
                }
                return true;
            case OBJECT:
                if (!(concreteNode instanceof ObjectNode)) {
                    errors.add(new MatchError(path, "an object node", "A "+concreteNode.getNodeType() + " node"));
                    return false;
                }
                return matchObject(path, (ObjectNode) matcherNode, (ObjectNode) concreteNode);
            case MISSING:
                if (!(concreteNode.getNodeType().equals(MISSING))) {
                    errors.add(new MatchError(path, "no node at all", "a node"));
                    return false;
                }
            case POJO:
            case BINARY:
                break;
        }
        errors.add(new MatchError(path, " a match implementation for " + matcherNode.getNodeType(), "Unimplemented"));
        return false;
    }

    private boolean matchPrimitive(JsonPath path, ValueNode matcherNode, ValueNode concreteNode) {
        String matcher = matcherNode.asText();
        String concrete = null;
        if (!(concreteNode instanceof NullNode)){
            concrete = concreteNode.asText();
        }
        boolean matches;
        if (matcher.startsWith("$")) {
            if (matcher.startsWith("$ID")) {
                JsonNode objRef = PathExtractor.extractPrimitiveNode(matcher,bindings);
                matches = matchPrimitive(path, (ValueNode)objRef, concreteNode);
            }
            else {
                try {
                    FieldMatcher<?> parseMatcher = parseMatcher(matcher);
                    matches = parseMatcher.matches(concrete, bindings);
                } catch (UnboundVariableException e) {
                    errors.add(new MatchError(path, e.getVar() + " to be bound", "unbound"));
                    return false;
                }
            }
        } else if (matcher.startsWith("\\$")) {
            String test = matcher.substring(1);
            return test.equals(concrete);
        } else {
            matches = matcher.equals(concrete);
        }

        if (!matches) {
            errors.add(new MatchError(path, matcher, concrete));
        }
        return matches;
    }

    public static FieldMatcher<?> parseMatcher(String matcher) {
        MatcherLexer l = new MatcherLexer(CharStreams.fromString(matcher));
        MatcherParser p = new MatcherParser(new CommonTokenStream(l));
        p.addErrorListener(new BaseErrorListener() {
            @Override
            public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine, String msg, RecognitionException e) {
                throw new IllegalStateException("failed to parse at line " + line + " due to " + msg, e);
            }
        });
        GrammarListenerMatcher listener = new GrammarListenerMatcher();
        p.addParseListener(listener);
        p.matcher();
        FieldMatcher<?> result = listener.result;
        if (result == null) {
            throw new UnsupportedOperationException("cant parse matcher " + matcher);
        }
        return result;
    }

    public final static Pattern sizePattern = Pattern.compile("([0-9]*)-([0-9]*)");

    private boolean matchArray(JsonPath path, ArrayNode matcherNode, ArrayNode concreteNode) {
        int matcherSize = matcherNode.size();
        int concreteSize = concreteNode.size();

        //special flags
        boolean strict = hasOption(MatcherOption.STRICT);
        boolean unordered = hasOption(MatcherOption.UNORDERED);
        boolean each = false;
        int min = 0;
        int max = Integer.MAX_VALUE;
        int offset = 0;
        boolean hasSize=false;

        if (matcherSize == 0) {
            if (concreteSize == 0) return true;
            else {
                errors.add(new MatchError(path, "and empty array", "size " + concreteSize));
            }
        }
        //look for the special matching node
        JsonNode n = matcherNode.get(0);
        if (n instanceof ObjectNode) {
            ObjectNode on = (ObjectNode) n;
            if (on.size() > 0) {
                if (on.fieldNames().next().startsWith("$")) {
                    //this is a "special" object
                    strict = on.has("$Strict");
                    unordered = on.has("$Unordered");
                    each = on.has("$Each");
                    JsonNode sizeNode = on.get("$Size");
                    if (sizeNode != null) {
                        hasSize = true;
                        String bounds = sizeNode.asText();
                        Matcher m = sizePattern.matcher(bounds);
                        if (m.matches()) {
                            if (m.group(1).length() > 0) {
                                min = Integer.parseInt(m.group(1));
                            }
                            if (m.group(2).length() > 0) {
                                max = Integer.parseInt(m.group(2));
                            }
                        } else {
                            errors.add(new MatchError(path, "size shoud be \"min-max\"", bounds));
                            return false;
                        }
                    }
                    if (strict || unordered || each || hasSize) {
                        offset = 1;
                        matcherSize--;
                    }
                }
            }
        }

        boolean matches = true;

        if (concreteSize < matcherSize) {
            errors.add(new MatchError(path, "an array of at least size " + matcherSize, Integer.toString(concreteSize)));
            return false;
        }

        if (strict && concreteSize > matcherSize) {
            errors.add(new MatchError(path, "an array of at exactly size " + matcherSize, Integer.toString(concreteSize)));
            matches = false;
        }

        if (concreteSize < min) {
            errors.add(new MatchError(path, "an array of at least size " + min, Integer.toString(concreteSize)));
            matches=false;
        }
        if (concreteSize > max) {
            errors.add(new MatchError(path, "an array of at most size " + max, Integer.toString(concreteSize)));
            matches = false;
        }

        if (!unordered) {
            for (int i = 0; i < matcherSize - offset; i++) {
                JsonNode matcherChild;
                if (each) {
                    matcherChild = matcherNode.get(offset);
                } else {
                    matcherChild = matcherNode.get(i + offset);
                }
                JsonNode concreteChild = concreteNode.get(i);
                matches &= matches(new JsonPath("[" + i + "]", path), matcherChild, concreteChild);
            }
        }
        else {
            List<JsonNode> leftToMatch = new ArrayList<>();
            concreteNode.elements().forEachRemaining(leftToMatch::add);
            //The test matchings here are going to make a mess of our errors list so lets temporarily replace it
            List<MatchError> realErrors = errors;
            errors = new ArrayList<>();
            for (int i = 0; i < matcherSize - offset; i++) {
                JsonNode matcherChild;
                if (each) {
                    matcherChild = matcherNode.get(offset);
                } else {
                    matcherChild = matcherNode.get(i + offset);
                }
                boolean isMatched=false;
                for (int j=0;j<leftToMatch.size();j++) {
                    isMatched = matches(new JsonPath("[" + i + "]", path), matcherChild, leftToMatch.get(j));
                    if (isMatched) {
                        leftToMatch.remove(j);
                        break;
                    }
                }
                if (!isMatched) {
                    realErrors.add(new MatchError(path, "an object matching "+matcherChild,"nothing matching"));
                }
                matches &= isMatched;
            }
            errors = realErrors;
        }
        return matches;
    }

    private boolean matchObject(JsonPath path, ObjectNode matcherNode, ObjectNode concreteNode) {
        boolean result = true;
        boolean strictMode = false;
        List<String> matchedKeys = new ArrayList<>();
        String id = null;
        for (Iterator<Map.Entry<String, JsonNode>> iterator = matcherNode.fields(); iterator.hasNext(); ) {
            Map.Entry<String, JsonNode> child = iterator.next();
            String key = child.getKey();

            Map<String, JsonNode> matchedNodes;
            if (key.startsWith("$")) {
                if (key.equals("$Strict")) {
                    strictMode = true;
                    continue;//ignore this
                }
                if (key.equals("$Size")) {
                    String bounds = child.getValue().asText();
                    Matcher m = sizePattern.matcher(bounds);
                    if (m.matches()) {
                        if (m.group(1).length() > 0) {
                            int min = Integer.parseInt(m.group(1));
                            if (concreteNode.size() > min) {
                                errors.add(new MatchError(path, "at least " + min + " keys", Integer.toString(concreteNode.size())));
                            }
                        }
                        if (m.group(2).length() > 0) {
                            int max = Integer.parseInt(m.group(2));
                            if (concreteNode.size() < max) {
                                errors.add(new MatchError(path, "at most " + max + " keys", Integer.toString(concreteNode.size())));
                            }
                        }
                    } else {
                        errors.add(new MatchError(path, "size shoud be \"min-max\"", bounds));
                        return false;
                    }
                }
                if (key.equals("$ID")) {
                    id = child.getValue().asText();
                    continue;
                }
                //interpret this node as a matcher
                FieldMatcher<?> matcher = parseMatcher(key);
                matchedNodes = new LinkedHashMap<>();
                for (Iterator<Entry<String, JsonNode>> citerator = concreteNode.fields(); citerator.hasNext(); ) {
                    Entry<String, JsonNode> cchild = citerator.next();
                    if (matcher.matches(cchild.getKey(), bindings)) {
                        matchedNodes.put(cchild.getKey(), cchild.getValue());
                    }
                }
            } else if (key.startsWith("\\$")) {
                key = key.substring(1);
                JsonNode matchedNode = concreteNode.get(key);
                if (matchedNode != null) {
                    matchedNodes = Map.of(key, matchedNode);
                } else {
                    matchedNodes = Map.of();
                }
            } else {
                JsonNode matchedNode = concreteNode.get(key);
                if (matchedNode != null) {
                    matchedNodes = Map.of(key, matchedNode);
                } else {
                    matchedNodes = Map.of();
                }
            }

            if (matchedNodes.isEmpty()) {
                errors.add(new MatchError(path, key, "not present"));
                return false;
            } else {
                matchedKeys.add(key);
                result &= matchedNodes.entrySet().stream()
                        .map(e -> matches(new JsonPath(e.getKey(), path),
                                child.getValue(),
                                e.getValue())).reduce(result, (r, b) -> r & b);
            }
        }
        if (strictMode) {
            List<String> concreteKeys = new ArrayList<>();
            concreteNode.fieldNames().forEachRemaining(concreteKeys::add);
            concreteKeys.removeAll(matchedKeys);
            if (!concreteKeys.isEmpty()) {
                errors.add(new MatchError(path, "no additional values", concreteKeys.toString()));
                result = false;
            }
        }

        if (result && id!=null) {
            bindings.put(id, concreteNode);
        }

        return result;

    }

    public List<MatchError> getErrors() {
        return errors;
    }

    public Map<String, Object> getBindings() {
        return bindings;
    }
}

File: messagematch-typescript/src/match/JsonMatcher.ts
import { readFileSync } from 'fs';
import MatcherLexer from '../antlr4ts/org/forwoods/messagematch/matchgrammar/MatcherLexer.js';
import MatcherParser from '../antlr4ts/org/forwoods/messagematch/matchgrammar/MatcherParser.js';
import GrammarListenerMatcher from './GrammarListenerMatcher';
import {FieldMatcher, UnboundVariableException} from "./fieldmatchers";

// Lightweight stand-ins for the Java classes used in original
export type JsonNode = any;

export class JsonPath {
	constructor(public pathElement: string, public parent: JsonPath | null) {}
	toString() { return this.parent ? `${this.parent}:${this.pathElement}` : this.pathElement; }
}

export class MatchError {
	constructor(public path: JsonPath, public expected: string, public actual: string) {}
	toString() { return `Error at ${this.path} expected matching ${this.expected} but was ${this.actual}`; }
}

export class JsonMatcher {
	static sizePattern = /([0-9]*)-([0-9]*)/;

	private errors: MatchError[] = [];
	private bindings: Map<string, any> = new Map();
	private options = new Set<string>();
	matchTime = -1;

	constructor(private matcherNode: JsonNode, private concreteNode: JsonNode) {}

	static readNodes(json: string): JsonNode {
		return JSON.parse(json);
	}

	hasOption(opt: string): boolean {
		return this.options.has(opt);
	}

	matches(): boolean {
		if (this.matchTime < 0) this.matchTime = Date.now();
		const i = this.matchTime;
		const d = new Date(i);
		this.bindings.set('date', d.toISOString().slice(0,10));
		this.bindings.set('time', d.toISOString().slice(11,19));
		this.bindings.set('datetime', this.matchTime);
		return this.matchesNode(new JsonPath('root', null), this.matcherNode, this.concreteNode);
	}

	private matchesNode(path: JsonPath, matcherNode: JsonNode, concreteNode: JsonNode): boolean {
		if (matcherNode === null) {
			if (concreteNode !== null) {
				this.errors.push(new MatchError(path, 'a null value', String(concreteNode)));
				return false;
			}
			return true;
		}
		if (typeof matcherNode === 'string' || typeof matcherNode === 'number' || typeof matcherNode === 'boolean') {
			if (concreteNode === null || concreteNode === undefined) {
				this.errors.push(new MatchError(path, 'a value node', String(concreteNode)));
				return false;
			}
			return this.matchPrimitive(path, matcherNode, concreteNode);
		}
		if (Array.isArray(matcherNode)) {
			if (!Array.isArray(concreteNode)) {
				this.errors.push(new MatchError(path, 'an array node', String(concreteNode)));
				return false;
			}
			return this.matchArray(path, matcherNode, concreteNode);
		}
		if (typeof matcherNode === 'object') {
			if (typeof concreteNode !== 'object' || Array.isArray(concreteNode)) {
				this.errors.push(new MatchError(path, 'an object node', String(concreteNode)));
				return false;
			}
			return this.matchObject(path, matcherNode, concreteNode);
		}
		this.errors.push(new MatchError(path, `a match implementation for ${typeof matcherNode}`, 'Unimplemented'));
		return false;
	}

	private matchPrimitive(path: JsonPath, matcher: any, concrete: any): boolean {
		const matcherStr = String(matcher);
		const concreteStr = concrete == null ? null : String(concrete);

		let matches = false;
		if (matcherStr.startsWith('$')) {
			if (matcherStr.startsWith('$ID')) {
				// TODO: implement path extractor; for now treat as fail
				matches = false;
			} else {
				try {
					const fm = this.parseMatcher(matcherStr);
					matches = fm.matches(concrete, this.bindings);
				} catch (e:any) {
					if (e instanceof UnboundVariableException) {}
					this.errors.push(new MatchError(path, (e as UnboundVariableException).varName + " to be bound", 'unbound'));
					return false;
				}
			}
		} else if (matcherStr.startsWith('\\$')) {
			const test = matcherStr.substring(1);
			matches = test === concreteStr;
		} else {
			matches = matcherStr === concreteStr;
		}

		if (!matches)
			this.errors.push(new MatchError(path, matcherStr, String(concrete)));
		return matches;
	}

	private parseMatcher(matcher: string): FieldMatcher<any> {
		// Use antlr4 generated lexer/parser to build a matcher listener
		const antlr4 = require('antlr4');
		const chars = new antlr4.InputStream(matcher);
		const lexer = new MatcherLexer(chars);
		const tokens = new antlr4.CommonTokenStream(lexer);
		const parser = new MatcherParser(tokens);
		// throw on syntax errors
		parser.removeErrorListeners();
		parser.addErrorListener({
			syntaxError(_recognizer:any, _offendingSymbol:any, line:number, _charPos:number, msg:string) {
				throw new Error(`failed to parse at line ${line} due to ${msg}`);
			}
		});
		// Use the GrammarListenerMatcher to build a FieldMatcher
		const listener = new GrammarListenerMatcher();

		parser.addParseListener(listener);

		parser.matcher();
		if (!listener.result) throw new Error('cant parse matcher ' + matcher);
		return listener.result;
	}

	private matchArray(path: JsonPath, matcherNode: JsonNode[], concreteNode: JsonNode[]): boolean {
		let matcherSize = matcherNode.length;
		const concreteSize = concreteNode.length;

		// special flags
		let strict = this.hasOption('$Strict');
		let unordered = this.hasOption('$Unordered');
		let each = false;
		let min = 0;
		let max = Number.MAX_SAFE_INTEGER;
		let offset = 0;
		let hasSize = false;

		if (matcherSize === 0) {
			if (concreteSize === 0) return true;
			this.errors.push(new MatchError(path, "an empty array", `size ${concreteSize}`));
			return false;
		}

		// look for special matching node
		const n = matcherNode[0];
		if (typeof n === 'object' && n !== null && !Array.isArray(n)) {
			const keys = Object.keys(n);
			if (keys.length > 0 && keys[0].startsWith('$')) {
				// this is a "special" object
				strict = '$Strict' in n;
				unordered = '$Unordered' in n;
				each = '$Each' in n;
				const sizeNode = n['$Size'];
				if (sizeNode !== undefined) {
					hasSize = true;
					const bounds = String(sizeNode);
					const m = JsonMatcher.sizePattern.exec(bounds);
					if (m) {
						if (m[1].length > 0) {
							min = parseInt(m[1], 10);
						}
						if (m[2].length > 0) {
							max = parseInt(m[2], 10);
						}
					} else {
						this.errors.push(new MatchError(path, "size should be \"min-max\"", bounds));
						return false;
					}
				}
				if (strict || unordered || each || hasSize) {
					offset = 1;
					matcherSize--;
				}
			}
		}

		let matches = true;

		if (concreteSize < matcherSize) {
			this.errors.push(new MatchError(path, `an array of at least size ${matcherSize}`, String(concreteSize)));
			return false;
		}

		if (strict && concreteSize > matcherSize) {
			this.errors.push(new MatchError(path, `an array of exactly size ${matcherSize}`, String(concreteSize)));
			matches = false;
		}

		if (concreteSize < min) {
			this.errors.push(new MatchError(path, `an array of at least size ${min}`, String(concreteSize)));
			matches = false;
		}
		if (concreteSize > max) {
			this.errors.push(new MatchError(path, `an array of at most size ${max}`, String(concreteSize)));
			matches = false;
		}

		if (!unordered) {
			for (let i = 0; i < matcherSize; i++) {
				const matcherChild = each ? matcherNode[offset] : matcherNode[i + offset];
				const concreteChild = concreteNode[i];
				matches = this.matchesNode(new JsonPath(`[${i}]`, path), matcherChild, concreteChild) && matches;
			}
		} else {
			const leftToMatch = [...concreteNode];
			// The test matchings here will mess up our errors list so temporarily replace it
			const realErrors = this.errors;
			this.errors = [];
			for (let i = 0; i < matcherSize; i++) {
				const matcherChild = each ? matcherNode[offset] : matcherNode[i + offset];
				let isMatched = false;
				for (let j = 0; j < leftToMatch.length; j++) {
					isMatched = this.matchesNode(new JsonPath(`[${i}]`, path), matcherChild, leftToMatch[j]);
					if (isMatched) {
						leftToMatch.splice(j, 1);
						break;
					}
				}
				if (!isMatched) {
					realErrors.push(new MatchError(path, `an object matching ${JSON.stringify(matcherChild)}`, "nothing matching"));
				}
				matches = isMatched && matches;
			}
			this.errors = realErrors;
		}
		return matches;
	}

	private matchObject(path: JsonPath, matcherNode: {[k:string]: any}, concreteNode: {[k:string]: any}): boolean {
		let result = true;
		const matchedKeys: string[] = [];
		let strictMode = false;
		let id: string|undefined;

		for (const key of Object.keys(matcherNode)) {
			const child = matcherNode[key];
			let matchedNodes: {[k:string]: any} = {};
			if (key.startsWith('$')) {
				if (key === '$Strict') {
					strictMode = true;
					continue;
				}
				if (key === '$Size') {
					const bounds = String(child);
					const m = JsonMatcher.sizePattern.exec(bounds);
					if (!m) { this.errors.push(new MatchError(path, 'size should be "min-max"', bounds)); return false; }
					throw new Error("Not implemented this");
					continue;
				}
				if (key === '$ID') { id = child; continue; }
				// matcher as key
				const matcher: FieldMatcher<any> = this.parseMatcher(key);
				matchedNodes = {};
				for (const k of Object.keys(concreteNode)) {
					if (matcher.matches(k, this.bindings)) {
						matchedNodes[k] = concreteNode[k];
					}
				}
			} else if (key.startsWith('\\$')) {
				const realKey = key.substring(1);
				if (realKey in concreteNode) matchedNodes[realKey] = concreteNode[realKey];
			} else {
				if (key in concreteNode) matchedNodes[key] = concreteNode[key];
			}

			if (Object.keys(matchedNodes).length === 0) {
				this.errors.push(new MatchError(path, key, 'not present'));
				return false;
			} else {
				matchedKeys.push(key);
				for (const [k, v] of Object.entries(matchedNodes)) {
					result = this.matchesNode(new JsonPath(k, path), child, v) && result;
				}
			}
		}

		if (strictMode) {
			let concreteKeys:string[] = []; 
			for (const key of Object.keys(concreteNode)) {
				concreteKeys.push(key);
			}
			
			concreteKeys = concreteKeys.filter(k=>!matchedKeys.includes(k));
			if (concreteKeys.length>0) {
				this.errors.push(new MatchError(path, 'no additional values', '['+concreteKeys.join(', ')+']'));
				result = false;
			}
		}

		if (result && id) this.bindings.set(id, concreteNode);
		return result;
	}

	getErrors(): MatchError[] { return this.errors; }
	getBindings(): Map<string, any> { return this.bindings; }
}
